### PR TITLE
TimeField with defaultValue persists time zone

### LINKED ISSUE #3606: TimePicker DateSegment format loses timezone when input values are removed
# 🐛 Bug Report 

@jshli and I found that removing parts of a time (eg. backspacing on the MM) causes the timezone to unset and revert to UTC.

## 🤔 Expected Behavior

The time and timezone should remain as it was before, aside from the removed parts reverting to placeholders.

## 😯 Current Behavior

The timezone disappears, and the time changes to the equivalent time in UTC. It does not return even if you fill in the missing segments again.

## 💁 Possible Solution

Allow the consumer to set a consistent timezone within the state.

## 🔦 Context

We have been trying to create a TimePicker component by passing in and returning Date objects (transforming between ZonedDateTime objects).

## 💻 Code Sample

You can see the behaviour on your docs (eg. https://react-spectrum.adobe.com/react-aria/useTimeField.html#granularity)

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | `"@react-stately/datepicker": "^3.0.2"`; `"@react-aria/datepicker": "^3.1.1"` |
| Browser          | Chrome |
| Operating System | macOS Monterey |

### PR DESCRIPTION
**Pull request by ktabors**
Closes #3606

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Goto the zoned story, delete the minute or hour segment.
The timeZone should remain.
Set a new value to the cleared segment.
The timeZone should remain.

## 🧢 Your Project:
RSP

### CODE REVIEW

File: packages/@react-stately/datepicker/src/useDateFieldState.ts [hunk 1]
```diff
@@ -156,16 +157,27 @@ export function useDateFieldState<T extends DateValue = DateValue>(props: DateFi
...
-  let [value, setDate] = useControlledState<DateValue>(
+  let [value, setDate2] = useControlledState<DateValue>(
     props.value,
     props.defaultValue,
     props.onChange
   );
 
-  let calendarValue = useMemo(() => convertValue(value, calendar), [value, calendar]);
+  let setDate = (v) => {
+    console.log('wrappingSetDate: clone, v', Object.assign({}, v), v);
```
Comment (ktabors) [inline packages/@react-stately/datepicker/src/useDateFieldState.ts:167]: trying to observe ever attempt to update `value`

File: packages/@react-stately/datepicker/src/useDateFieldState.ts [hunk 2]
```diff
@@ -156,16 +157,27 @@ export function useDateFieldState<T extends DateValue = DateValue>(props: DateFi
...
+  }
+
+  let defaultFormatter = useMemo(() => new DateFormatter(locale), [locale]);
+  let calendar = useMemo(() => {
+    console.log('useMemo: defaultFormatter.resolvedOptions(), value', defaultFormatter.resolvedOptions());
+    return createCalendar(defaultFormatter.resolvedOptions().calendar);
+  }, [createCalendar, defaultFormatter]);
+
+  let calendarValue = useMemo(() => {
+    console.log('useMemo: value, calendar, convert', value, calendar, convertValue(value, calendar));
```
Comment (ktabors) [inline packages/@react-stately/datepicker/src/useDateFieldState.ts:178]: After `value` is updated I wanted to see how it was mutated, but found this was getting an unexpected value.

Also this and the useMemo above get called a few times each after a value update and I'd only expect once.

File: packages/@react-stately/datepicker/src/useDateFieldState.ts [hunk 3]
```diff
@@ -287,6 +304,7 @@ export function useDateFieldState<T extends DateValue = DateValue>(props: DateFi
   };
 
   let adjustSegment = (type: Intl.DateTimeFormatPartTypes, amount: number) => {
+    console.log('who is calling adjustSegment?');
```
Comment (ktabors) [inline packages/@react-stately/datepicker/src/useDateFieldState.ts:307]: At one point I was looking at segments as a potential source of the issue.

File: packages/@react-stately/datepicker/src/useTimeFieldState.ts [hunk 4]
```diff
@@ -54,7 +60,7 @@ export function useTimeFieldState<T extends TimeValue = TimeValue>(props: TimeFi
   return useDateFieldState({
     ...props,
     value: dateTime,
-    defaultValue: undefined,
+    defaultValue: convertValue(props.defaultValue),
```
Comment (ktabors) [inline packages/@react-stately/datepicker/src/useTimeFieldState.ts:63]: Delete

File: packages/@react-stately/datepicker/src/useTimeFieldState.ts [hunk 5]
```diff
@@ -42,7 +42,13 @@ export function useTimeFieldState<T extends TimeValue = TimeValue>(props: TimeFi
 
   let v = value || placeholderValue;
   let day = v && 'day' in v ? v : undefined;
-  let placeholderDate = useMemo(() => convertValue(placeholderValue), [placeholderValue]);
+  let placeholderDate = useMemo(() => {
+    let valueTimeZone = v && 'timeZone' in v ? v.timeZone : undefined;
+    let defaultValueTimeZone = props.defaultValue && 'timeZone' in props.defaultValue ? props.defaultValue.timeZone : undefined;
+
+    //console.log('placeholder', valueTimeZone || defaultValueTimeZone ? toZoned(convertValue(placeholderValue), valueTimeZone || defaultValueTimeZone) : convertValue(placeholderValue));
+    return valueTimeZone || defaultValueTimeZone ? toZoned(convertValue(placeholderValue), valueTimeZone || defaultValueTimeZone) : convertValue(placeholderValue);
```
Comment (ktabors) [inline packages/@react-stately/datepicker/src/useTimeFieldState.ts:50]: This is the fix for the bug as reported, but there is an additional issue(s?) and it isn't the complete solution.

File: packages/@react-stately/utils/src/useControlledState.ts [hunk 6]
```diff
@@ -65,10 +67,12 @@ export function useControlledState<T>(
       setStateValue(updateFunction);
     } else {
       if (!isControlled) {
+        console.log('setValue: setting state clone and value', Object.assign({}, value), value);
         setStateValue(value);
       }
       onChangeCaller(value, ...args);
     }
+    console.log('did I end up here?', isControlled, Object.assign({}, value));
```
Comment (ktabors) [inline packages/@react-stately/utils/src/useControlledState.ts:75]: These show the setDate/value update I expected, but it is called by something I'm not able to track to update with a different value at the same time. Late night thought that we could have two instances of the state object?

**Review by reidbarber — APPROVED** [review]
LGTM

**Review by snowystinger — APPROVED** [review]
lgtm

**Review by devongovett — APPROVED** [review]

**Pull request merged**